src/backend/env/auction.rs

```diff
@@ -0,0 +1,393 @@
+        }
+        let mut amount = self.amount;
+        let mut bids = Vec::default();
+
+        while amount > 0 {
+            let Some(bid) = self.bids.pop_last() else {
+                break;
+            };
+            let eff_amount = bid.amount.min(amount);
+            bids.push(Bid {
+                user: bid.user,
+                amount: eff_amount,
+                e8s_per_token: bid.e8s_per_token,
+                timestamp: bid.timestamp,
+            });
+            // if the bid was larger than tokens left, create a leftover bid
+            if eff_amount < bid.amount {
+                self.bids.insert(Bid {
+                    user: bid.user,
+                    amount: bid.amount - eff_amount,
+                    e8s_per_token: bid.e8s_per_token,
+                    timestamp: bid.timestamp,
+                });
+            }
+            amount = amount.saturating_sub(eff_amount);
+        }
+
+        bids
+    }
+}
+
+/// Cancels user's bid and returns the funds to user wallet.
+pub async fn cancel_bid(principal: Principal) -> Result<u64, String> {
+    let bid = mutate(|state| remove_bid(state, principal))?;
+
+    let user_account = read(|state| {
+        let user_principal = state
+            .principal_to_user(principal)
+            .expect("no user found")
+            .principal;
+        AccountIdentifier::new(&user_principal, &DEFAULT_SUBACCOUNT)
+    });
+
+    let funds = bid
+        .amount
+        .checked_mul(bid.e8s_per_token)
+        .expect("overflow")
+        .checked_sub(DEFAULT_FEE.e8s())
+        .expect("nothing to refund");
+
+    invoices::transfer(
+        user_account,
+        Tokens::from_e8s(funds),
+        Memo(727),
+        Some(Subaccount(AUCTION_ICP_SUBACCOUNT)),
+    )
+    .await
+    .map_err(|err| {
+        let msg = format!("couldn't withdraw funds from bid {:?}: {}", bid, err);
+        mutate(|state| state.logger.error(&msg));
+        msg
+    })
+}
+
+fn remove_bid(state: &mut State, principal: Principal) -> Result<Bid, String> {
+    let user_id = state
+        .principal_to_user(principal)
+        .ok_or("no user found")?
+        .id;
+
+    let bid = state
+        .auction
+        .bids
+        .iter()
+        .find(|bid| bid.user == user_id)
+        .ok_or("no bids found")?
+        .clone();
+    state.auction.bids.retain(|bid| bid.user != user_id);
+
+    Ok(bid)
+}
+
+fn has_bid(state: &State, principal: Principal) -> bool {
+    state
+        .principal_to_user(principal)
+        .map(|user| state.auction.bids.iter().any(|bid| bid.user == user.id))
+        .unwrap_or_default()
+}
+
+/// Creates a new user bid. Requires a transfer to the user subaccount before.
+pub async fn create_bid(
+    principal: Principal,
+    amount: Token,
+    e8s_per_token: u64,
+) -> Result<(), String> {
+    // cancel existing bid if necessary
+    if read(|state| has_bid(state, principal)) {
+        cancel_bid(principal).await?;
+    }
+
```

Comment:
Maybe we should `assert!(!has_bid(state, principal));` here and after the transfer to avoid any kind of potential races?

Reply:
I see no good reason for this. This behaviour can only be triggered by a broken client and the second assert will still not prevent the "loss" of funds, right?